File: erlang/list-ops/src/list_ops.erl
-module(list_ops).

-export([append/2, concat/1, filter/2, length/1, map/2, foldl/3, foldr/3, reverse/1]).

append([], List) -> List;
append([X | Xs], List) -> [X | append(Xs, List)].

concat([]) -> [];
concat([X | Xs]) -> append(X, concat(Xs)).

filter(_, []) -> [];
filter(F, [X | Xs]) ->
    case F(X) of
      true -> [X | filter(F, Xs)];
      _ -> filter(F, Xs)
    end.

length([]) -> 0;
length([_ | Xs]) -> 1 + list_ops:length(Xs).

map(_, []) -> [];
map(F, [X | Xs]) ->
    [F(X) | map(F, Xs)].

foldl(_, Acc, []) -> Acc;
foldl(F, Acc, [X | Xs]) ->
    foldl(F, F(X, Acc), Xs).

foldr(_, Acc, []) -> Acc;
foldr(F, Acc, [X | Xs]) ->
    F(X, foldr(F, Acc, Xs)).

reverse(List) -> reverse(List, []).
reverse([], Acc) -> Acc;
reverse([X | Xs], Acc) -> reverse(Xs, [X | Acc]).

File: php/list-ops/ListOps.php
<?php

class ListOps
{
    public function append(array $list1, array $list2): array
    {
        return [...$list1, ...$list2];
    }
    public function concat(array $list1, array ...$lstn): array
    {
        return array_reduce($lstn, fn ($acc, $lst) => [...$acc, ...$lst], $list1);
    }
    public function filter(callable $f, array $lst): array
    {
        return array_values(array_filter($lst, $f));
    }
    public function length(array $lst): int
    {
        return count($lst);
    }
    public function map(callable $f, array $lst): array
    {
        return array_map($f, $lst);
    }
    public function foldl(callable $f, array $lst, $acc)
    {
        return array_reduce($lst, $f, $acc);
    }
    public function foldr(callable $f, array $lst, $acc)
    {
        return array_reduce($this->reverse($lst), $f, $acc);
    }
    public function reverse(array $lst): array
    {
        return array_reverse($lst);
    }
}

